### ISSUE TITLE
Fully convolutional policies?

### ISSUE THREAD

**Issue by dwilson1988** [reinforceio/tensorforce#464]
Is there a way to create a fully convolutional policy? It seems like since the output layers are added, this wouldn't be possible. Is there a way to do this with a custom network?

**Comment by AlexKuhnle** [comment]
What would be a fully convolutional policy? As far as I know, fully convolutional nets just use convolutions everywhere, since dense layers are a 'special case' of convolutions... If that's so, it shouldn't make a difference that the output layers are instantiated as linear layers instead of convolutions acting as such. But maybe I'm missing something here...

**Comment by dwilson1988** [comment]
My particular use case currently, and I'm sure there are others as well, is to select a spatial action over a NxN grid. One major advantage of a fully convolutional network is it can be applied to various size inputs, so, for example I could train my policy network over a small chip, but evaluate the same policy over a larger image. Another advantage is the number of parameters can potentially be much smaller since there are no dense layers. For my use case, the last layer would be a conv2d of shape 1x1x1 to collapse the final feature map into a single image, then run through a softmax activation, where I could sample the spatial action. (See pg. 10 of DeepMind/Blizzard's starcraft 2 learning environment paper, https://deepmind.com/documents/110/sc2le.pdf)

**Comment by AlexKuhnle** [comment]
Understood. Hmm, you could implement this as a separate `Distribution` class, more or less in parallel to the existing `Categorical` distribution, but just accepting whatever input shape and doing a categorical selection over it. What do you think?

**Comment by dwilson1988** [comment]
That makes sense. The Distribution class defines the final output layer for that action, yes? How would I then differentiate between which Distribution would be used in the action space?

**Comment by AlexKuhnle** [comment]
The `Distribution` class usually defines necessary final linear transformations, yes. You can specify a different distribution via `distributions_spec`, e.g. `action1="custom"` (if it's [integrated](https://github.com/reinforceio/tensorforce/blob/master/tensorforce/core/distributions/__init__.py#L23) into the TensorForce library) or `action1="mymodule.MyClass"` or so if not. Since it sounds like a generally useful feature, I think it would be great if you would contribute it once it's working. :-)

**Comment by dwilson1988** [comment]
Awesome! That sounds like what I'm looking for. Since I will likely need this for something I'm working on, I'll attempt to implement and send a PR when I'm happy with it. Thank you!

**Comment by AlexKuhnle** [comment]
No worries, and happy to be of help (best via Gitter) if needed.

**Comment by jerabaul29** [comment]
Has there been some follow up of this? Any code pushed? ;)

**Comment by dwilson1988** [comment]
@jerabaul29, I've gone a different direction with my observation space, so the learning from pixels side is no longer needed. I'd love to help out, I just haven't had the time.

**Comment by jerabaul29** [comment]
No worries, just asking by curiosity ;) I also think I will prefer another approach, but asking in case this would have been easy to test.

**Comment by AlexKuhnle** [comment]
Feel free to reopen if you're interested in it. I don't think it's an actually problematic feature, at least to implement it (wouldn't know how to evaluate it straight away).
